### 10.数学建模综合案例分析：投资组合风险分析

### 10.数学建模综合案例分析：投资组合风险分析

<div class="alert alert-warning" role="alert">
  <h3>📋任务</h3> 
</div>

GitModel公司是一家专业的投资银行，志在帮助客户更好地管理资产。客户手头上有一笔100万的资金，希望将这笔钱投入股票市场进行投资理财，投资人看中了两个股票$A$、$B$，股票分析师通过对股票$A$、$B$的历史数据分析发现：股票$A$的平均收益近似服从$N(0.1,0.01)$，股票B的平均收益近似服从$N(0.3,0.04)$。现在客户希望通过分析得出投资股票$A$、$B$的最佳组合（在预期收益确定情况下最小风险时，需要投资$A$、$B$的份额）。

分析：

首先，我们先来分析投资组合的收益应该如何计算：设$A$、$B$的投资收益率为随机变量$X$、$Y$，因此$X～N(0.1,0.01)$，$Y～N(0.3,0.04)$。设$x_1$为投资A的份额，$y_1=1-x_1$为投资B的份额，因此投资组合的收益率为：$Z = x_1*X + y_1*Y$，投资组合的平均收益率为：$E(Z) = x_1*E(X) + y_1*E(Y)$。

接下来，我们来分析投资组合的风险应该如何计算：何为风险，最简单来说就是收益的不确定性，如果收益是确定且固定的，就无所谓的风险可言。根据对风险的直观描述，我们可以定义风险为收益率的方差，因此：股票A的风险为$\sigma_x^2 = 0.01$，股票B的风险为$\sigma_y^2 = 0.04$,而投资组合的风险为
$$
\begin{aligned}
Var(Z) &= Var(x_1*X + y_1*Y)\\
&=x_{1}^{2} \operatorname{Var}(X)+y_{1}^{2} \operatorname{Var}(Y)+2 x_{1}y_{1} \operatorname{Cov}(X, Y)
\end{aligned}
$$
因此，最佳的投资组合应该是风险最小时的投资组合，即：
$$
\begin{aligned}
&min \quad Var(Z) \\
&= min \quad x_{1}^{2} \operatorname{Var}(X)+y_{1}^{2} \operatorname{Var}(Y)+2 x_{1}y_{1} \operatorname{Cov}(X, Y)\\
&=\frac{d(Var(Z))}{d(x_1)} = 0
\end{aligned}
$$

In [23]:
from sympy import *
from scipy.stats import norm

# 投资份额,它们的相关系数
x = symbols('x')
y = symbols('y')
corr_xy = symbols('corr_xy')
Cov_xy = symbols('Cov_xy')
y = 1-x


# 随机变量
X = norm(loc=0.1,scale=0.1)
Y = norm(loc=0.3,scale=0.2)
Dx, Dy = X.var(), Y.var()

In [26]:
## 请根据var_Z的定义写出相应的公式代码
Cov_xy = corr_xy*sqrt(Dx*Dy)
var_z = (x**2*Dx + y**2*Dy + 2*x*y*Cov_xy)

# 计算最小风险的投资组合
def min_risk_group(var_z,corr_xy):
    # 求驻点
    var_z_df1 = diff(var_z,x)
    x_value = solve(var_z_df1)
    # 二阶导在该点的值>0为极小值
    var_z_df2 = diff(var_z_df1,x)
    x_value2 = var_z_df2.evalf(subs={x:x_value[0]})
    
    if x_value2 < 0:
        x_value.delete(x_value[0])
    # 极小值的最小值者
    var_z_min  = var_z.evalf(subs={x:x_value[0]})
    return   x_value[0],var_z_min

In [27]:
## 请根据相应的优化分析写出代码
# 寻找最佳投资组合
corr_xy = 0.4   # 假设corr_xy = 0.4

x,var_z_min = min_risk_group(var_z,corr_xy)
y = 1-x
# 判断相关系数对方案份额分配有无实际意义
if (x > 1)  or (x < 0):
    print("当相关系数corr_xy为{:.2f}时，投资方案无意义。".format(corr_xy))
else:
    print("因此，根据风险最小化原则，应该投资A股票：100w * {}, 而投资B股票：100w *{}".format(x, y))

因此，根据风险最小化原则，应该投资A股票：100w * 0.941176470588235, 而投资B股票：100w *0.0588235294117647


因此，根据风险最小化原则，应该投资A股票$10w \times 0.94$元，而投资B股票$10w \times 0.06$元。
<div class="alert alert-success" role="alert">
  <h4 class="alert-heading">Well done!</h4>
  <p>恭喜你，完成了《动手学概率论》的学习，希望在日后的数模竞赛中，通过本课程的学习，你可以更加灵活分析不确定性世界，取得优异的成绩.</p>
  <hr>
</div>